File: tutorials/007_read_detection_boxes/read_detection_boxes.py
#!/usr/bin/env python3
"""Decode YOLO detections through ModelOptions and read the BBOX tensor.

The output tensor is a rank-1 uint8 buffer: a uint32 count header followed by
N 24-byte RawBox records (int32 x, y, w, h; float32 score; int32 class_id).

Usage:
  python3 read_detection_boxes.py --model /path/to/yolo_v8s.tar.gz --image /path/to.jpg
"""
from __future__ import annotations

import argparse
import struct
import sys
from pathlib import Path

try:
  import cv2
  import pyneat
except ImportError:
  sys.exit(
      "pyneat is not importable. Either Neat is not installed, or the venv is not activated.\n"
      "Run: source ~/pyneat/bin/activate"
  )

def main(argv: list[str]) -> int:
  ap = argparse.ArgumentParser(description=__doc__)
  ap.add_argument("--model", type=Path, required=True)
  ap.add_argument("--image", type=Path, required=True)
  args = ap.parse_args(argv[1:])

  bgr = cv2.imread(str(args.image), cv2.IMREAD_COLOR)
  if bgr is None:
    raise RuntimeError(f"failed to load image: {args.image}")
  height, width, channels = bgr.shape

  # STEP configure-decode
  opt = pyneat.ModelOptions()
  opt.preprocess.kind = pyneat.InputKind.Image
  opt.preprocess.enable = pyneat.AutoFlag.On
  opt.preprocess.color_convert.input_format = pyneat.PreprocessColorFormat.BGR
  opt.preprocess.input_max_width = width
  opt.preprocess.input_max_height = height
  opt.preprocess.input_max_depth = channels
  opt.preprocess.normalize.enable = pyneat.AutoFlag.On
  opt.decode_type = pyneat.BoxDecodeType.YoloV8
  opt.score_threshold = 0.55
  opt.nms_iou_threshold = 0.50
  opt.top_k = 100
  opt.boxdecode_original_width = width
  opt.boxdecode_original_height = height
  # END STEP

  # CORE LOGIC
  # STEP load-model
  model = pyneat.Model(str(args.model), opt)
  # END STEP

  # STEP run-decode
  tensor = pyneat.Tensor.from_numpy(bgr, copy=True, image_format=pyneat.PixelFormat.BGR)
  outputs = model.run([tensor], timeout_ms=30000)
  if not outputs:
    raise RuntimeError("model produced no outputs")
  # END STEP
  # END CORE LOGIC

  # STEP read-boxes
  # Two paths for reading the output:
  #   - Runtimes that wire BoxDecode into model.run produce one BBOX uint8 tensor.
  #   - Runtimes that do not return the raw MLA feature-map tensors instead.
  # The shipped README explains the BBOX wire format (uint32 count + N 24-byte RawBox).
  first = outputs[0] if outputs else None
  detection = getattr(first.semantic, "detection", None) if first is not None else None
  if first is not None and detection is not None and getattr(detection, "format", "") == "BBOX":
    buf = bytes(first.to_numpy(copy=False))
    detections = struct.unpack_from("<I", buf, 0)[0] if len(buf) >= 4 else 0
    print(f"detections={detections}")
    for index in range(detections):
      offset = 4 + index * 24
      if offset + 24 > len(buf):
        raise RuntimeError("truncated BBOX output")
      x, y, w, h, score, class_id = struct.unpack_from("<iiiifi", buf, offset)
      print(
        f"box[{index}] class={class_id} score={score:.6f} "
        f"xyxy=[{x},{y},{x + w},{y + h}]"
      )
    if detections == 0:
      raise RuntimeError("image produced no detections above threshold")
  else:
    raise RuntimeError(f"BoxDecode not wired by this runtime; raw_output_heads={len(outputs)}")
  # END STEP
  return 0


if __name__ == "__main__":
  raise SystemExit(main(sys.argv))

File: tutorials/007_read_detection_boxes/read_detection_boxes.cpp
// Decompose model execution into stages: Preproc -> Infer -> BoxDecode.
//
// Usage:
//   tutorial_007_read_detection_boxes --model /path/to/yolo_v8s.tar.gz --image /path/to.jpg

#include "neat.h"

#include "pipeline/StageRun.h"

#include <opencv2/imgcodecs.hpp>

#include <iostream>
#include <stdexcept>
#include <string>

namespace {

bool get_arg(int argc, char** argv, const std::string& key, std::string& out) {
  for (int i = 1; i + 1 < argc; ++i) {
    if (key == argv[i]) {
      out = argv[i + 1];
      return true;
    }
  }
  return false;
}

} // namespace

int main(int argc, char** argv) {
  try {
    std::string model_path, image;
    if (!get_arg(argc, argv, "--model", model_path) || !get_arg(argc, argv, "--image", image)) {
      std::cerr << "Usage: tutorial_007_read_detection_boxes --model <path> --image <path>\n";
      return 1;
    }

    cv::Mat bgr = cv::imread(image, cv::IMREAD_COLOR);
    if (bgr.empty())
      throw std::runtime_error("failed to load image: " + image);

    // STEP configure-decode
    simaai::neat::Model::Options opt;
    opt.preprocess.kind = simaai::neat::InputKind::Image;
    opt.preprocess.enable = simaai::neat::AutoFlag::On;
    opt.preprocess.color_convert.input_format = simaai::neat::PreprocessColorFormat::BGR;
    opt.preprocess.input_max_width = bgr.cols;
    opt.preprocess.input_max_height = bgr.rows;
    opt.preprocess.input_max_depth = bgr.channels();
    opt.preprocess.normalize.enable = simaai::neat::AutoFlag::On;
    opt.decode_type = simaai::neat::BoxDecodeType::YoloV8;
    // END STEP

    // STEP load-model
    simaai::neat::Model model(model_path, opt);
    // END STEP

    // CORE LOGIC
    // Stage-by-stage: each stages::* call runs one piece of the model pipeline.
    // STEP run-decode
    simaai::neat::TensorList pre = simaai::neat::stages::Preproc(std::vector<cv::Mat>{bgr}, model);
    simaai::neat::Sample infer_samples = simaai::neat::stages::Infer(
        simaai::neat::Sample{simaai::neat::sample_from_tensors(pre)}, model);
    if (infer_samples.empty())
      throw std::runtime_error("infer stage returned no samples");
    simaai::neat::Sample infer = infer_samples.front();

    simaai::neat::stages::BoxDecodeOptions box(simaai::neat::BoxDecodeType::YoloV8);
    (void)box.decode_type;
    (void)bgr.cols;
    (void)bgr.rows;
    box.detection_threshold = 0.55;
    box.nms_iou_threshold = 0.5;
    box.top_k = 100;
    // END STEP

    // STEP read-boxes
    // BoxDecode parses the "BBOX" tensor into {x1, y1, x2, y2, score, class_id}
    // entries clamped to original_width x original_height source pixels.
    simaai::neat::BoxDecodeResultList decoded_results =
        simaai::neat::stages::BoxDecodeResults(simaai::neat::Sample{infer}, model, box);
    if (decoded_results.empty())
      throw std::runtime_error("boxdecode result parser returned no results");
    const simaai::neat::BoxDecodeResult& decoded = decoded_results.front();
    // END STEP
    // END CORE LOGIC

    std::cout << "boxes=" << decoded.boxes.size() << "\n";
    for (std::size_t i = 0; i < decoded.boxes.size(); ++i) {
      const auto& detection = decoded.boxes[i];
      std::cout << "box[" << i << "] class=" << detection.class_id
                << " score=" << detection.score << " xyxy=[" << detection.x1 << ","
                << detection.y1 << "," << detection.x2 << "," << detection.y2 << "]\n";
    }
    if (decoded.boxes.empty())
      throw std::runtime_error("image produced no detections above threshold");
    std::cout << "[OK] 007_read_detection_boxes\n";
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "[FAIL] " << e.what() << "\n";
    return 1;
  }
}
